dma/sniff_crc/sniff_crc.c

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) 2023 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+// Use the DMA engine's 'sniff' capability to calculate a CRC32 on data copied between two buffers in memory.
+
+#include <stdio.h>
+#include <string.h>
+#include "pico/stdlib.h"
+#include "hardware/dma.h"
+
+#define CRC32_INIT                  ((uint32_t)-1l)
+
+#define DATA_TO_CHECK_LEN           9
+#define CRC32_LEN                   4
+#define TOTAL_LEN                   (DATA_TO_CHECK_LEN + CRC32_LEN)
+
+// commonly used crc test data and also space for the crc
+static uint8_t src[TOTAL_LEN] = { 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x00, 0x00, 0x00, 0x00 };
+static uint8_t dst[TOTAL_LEN];
+
+// This uses the "reversed" polynomial and shift direction
+static uint32_t soft_crc32_block(uint32_t crc, uint8_t *bytp, uint32_t length) {
+    while(length--) {
+        uint32_t byte32 = (uint32_t)*bytp++;
+
+        for (uint8_t bit = 8; bit; bit--, byte32 >>= 1) {
+            crc = (crc >> 1) ^ (((crc ^ byte32) & 1ul) ? 0xEDB88320ul : 0ul);
+        }
+    }
+    return crc;
+}
+
+int main() {
+    uint32_t crc_res;
+
+    stdio_init_all();
+
+    printf("\n");
+    printf("\n");
+
+    // cacluate and append the crc
+    crc_res = soft_crc32_block(CRC32_INIT, src, DATA_TO_CHECK_LEN);
+    *((uint32_t *)&src[DATA_TO_CHECK_LEN]) = crc_res;
+
+    printf("Buffer to DMA: ");
+    for (int i = 0; i < 12; i++) {
```

Comment:
magic number 12. Not obvious why 12 is used here either.

Reply:
Thanks, that crept through.  It will be TOTAL_LEN